### ISSUE TITLE
GPU memory utilization and options

### ISSUE THREAD

**Issue by wswdx** [NVIDIA/vid2vid#65]
I have a few questions about memory utilization during training **pose2body** with 8 GTX1080ti:
1. The original setting of your code is: 
...
--gpu_ids 0,1,2,3,4,5,6,7 --batchSize 8 --max_frames_per_gpu 3 \
--niter 5 --niter_decay 5 \
--no_first_img --n_frames_total 12 --max_t_step **4**
which cause the "Out of memory" error, so I modified the below options like this:
**--batchSize 1 --max_frames_per_gpu 1 --n_gpus_gen 6**\
and the "Out of memory" error disappeared. But the batch size decreases a lot, can I still reach the same performance as the batch size set to 8 (how about training for more epochs)? 
2. When I tried to set the batch size to 2, I also get the "Out of memory" error. Does the batch size can only be set to 1 or "n_gpus_gen" as you mentioned in "more training/test details" of ReadMe? If so, how to set "n_gpus_gen"? Is "n_gpus_gen=6" OK for pose2body training?
3. When I use the options as showed in **1**, I find that memory of the first GPU(id=0) was occupied much more than other gpus while no others use the gpus at the same time. Why? How can I make the memory occupied to be closed among 8 GPUs? I think it maybe a reason of why I can't use bigger batch size.
Please help me out of these questions, I will appreciate your help very much! Thank you!

**Comment by tcwang0509** [comment]
1. Yes currently only 1 or n_gpus_gen is supported. In your case, if you set `n_gpus_gen=6` and `batchSize=2`, the network will just try to distribute the the two batches onto the first 2 of the 6 gpus, which is apparently inefficient. I think batchSize=1 and n_gpus_gen=6 is fine, just probably need to train longer.
2. As explained above, batchSize=2 is not efficient yet... sorry for the inconvenience.

**Comment by wswdx** [comment]
> 1. Yes currently only 1 or n_gpus_gen is supported. In your case, if you set `n_gpus_gen=6` and `batchSize=2`, the network will just try to distribute the the two batches onto the first 2 of the 6 gpus, which is apparently inefficient. I think batchSize=1 and n_gpus_gen=6 is fine, just probably need to train longer.
> 2. As explained above, batchSize=2 is not efficient yet... sorry for the inconvenience.

Thank you for your reply, I ! I have one more question,  when the batch size = 1, memory of the 1st GPU was occupied much more than the other 7 GPUs, and why? Is that because when batch size = 1, the generator network will only be placed at the 1st GPU? 
Hope you can help me, it bothers me much, and thank you in advance!

**Comment by tcwang0509** [comment]
There is always overhead for the first GPU during multi-GPU training (assigning inputs, gather gradients, etc). This is most likely inherent to pytorch and there's little I can do.

**Comment by wswdx** [comment]
>           There is always overhead for the first GPU during multi-GPU training (assigning inputs, gather gradients, etc). This is most likely inherent to pytorch and there's little I can do.

Thank you! Understood. I never used pytorch before, that's why I was confused. Thanks for your explanation.
